otherlibs/unix/spawn.c

```diff
@@ -0,0 +1,138 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*   Xavier Leroy, projet Cambium, Collège de France and INRIA Paris      */
+/*                                                                        */
+/*   Copyright 2020 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define _GNU_SOURCE  /* helps to find execvpe() */
+#include <errno.h>
+#include <sys/types.h>
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include "unixsupport.h"
+
+#ifdef HAS_POSIX_SPAWN
+
+#include <spawn.h>
+
+extern char ** environ;
+
+CAMLprim value unix_spawn(value executable,
+                          value args,
+                          value optenv,
+                          value usepath,
+                          value redirect)
+{
+  char ** argv;
+  char ** envp;
+  const char * path;
+  pid_t pid;
+  int src, dst, r;
+  posix_spawn_file_actions_t act;
+
+  caml_unix_check_path(executable, "create_process");
+  path = String_val(executable);
+  argv = cstringvect(args, "create_process");
+  if (Is_block(optenv)) {
+    envp = cstringvect(Field(optenv, 0), "create_process");
+  } else {
+    envp = environ;
+  }
+  posix_spawn_file_actions_init(&act);
+  for (dst = 0; dst <= 2; dst++) {
+    src = Int_val(Field(redirect, dst));
+    if (src != dst) {
+      r = posix_spawn_file_actions_adddup2(&act, src, dst);
+      if (r != 0) goto error;
+      r = posix_spawn_file_actions_addclose(&act, src);
+      if (r != 0) goto error;
```

Comment:
This is a minor point, but I think it would be safer to perform all `dup2` actions before all `close` actions, in case two of the source descriptors (typicall `stdout` and `stderr`) are themselves duplicates.

Reply:
Well spotted!  Indeed, the case `>file 2>&1` was handled incorrectly.  It's not enough to do all the close at the end, because we end up closing the same descriptor twice.

Commit c5d9ab31c fixes the issue and commit 41204826b adds a test.

Reply:
Thanks! I reviewed the fix and I believe that (while it does not simplify the control flow of this piece of code) it is correct.

Reply:
I know that's a lot of gotos, but for a good cause.